File: src/main/java/com/vc/hard/L815.scala
package com.vc.hard

object L815 {
  def numBusesToDestination(routes: Array[Array[Int]], S: Int, T: Int): Int = {
    import scala.collection.mutable
    /**
        [
          [1, 2, 7],
          [3, 6, 7]
        ]
    */
    if(S == T) return 0
    val map = new mutable.HashMap[Int, mutable.HashSet[Int]]()
    routes.indices.foreach(i => {
      routes(i).indices.foreach(j => {
        val set = map.getOrElse(routes(i)(j), new mutable.HashSet[Int]())
        set += i
        map.put(routes(i)(j), set)
      })
    })

    /**
      Map(
        2 -> Set(0),
        7 -> Set(0, 1),
        1 -> Set(0),
        3 -> Set(1),
        6 -> Set(1)
      )
    */
    val q = new mutable.Queue[Int]()
    val visited = new mutable.HashSet[Int]()
    q.enqueue(S)
    visited += S
    var res = 0
    while(q.nonEmpty) {
      val size = q.size
      res += 1
      var i = 0
      while(i < size) {
        val e = q.dequeue()
        if(map.contains(e)) {
          map(e).foreach(index => {
            routes(index).foreach(child => {
              if(!visited(child)) {
                visited += child
                if(child == T) return res
                q += child
              }
            })
          })
        }
        i += 1
      }
    }
    -1
  }
}

File: src/main/java/com/vc/hard/L815.java
package com.vc.hard;

import java.util.*;

class L815 {
    public int numBusesToDestination(int[][] routes, int S, int T) {
        /**
             [1, 2, 7]
             [3, 6, 7]

             1 -> 0
             2 -> 0
             7 -> 0, 1
             3 -> 1
             6 -> 1
             7 -> 1
        */
        if(S == T) return 0;
        HashMap<Integer, HashSet<Integer>> stations = new HashMap<>();
        for(int i = 0; i < routes.length; i++) {
            for(int j = 0; j < routes[i].length; j++) {
                int station = routes[i][j];
                HashSet<Integer> buses = stations.getOrDefault(station, new HashSet<Integer>());
                buses.add(i);
                stations.put(station, buses);
            }
        }
        //System.out.println(stations);

        //BFS
        Queue<Integer> q = new LinkedList<>();
        HashSet<Integer> stationVisited = new HashSet<>();
        HashSet<Integer> busVisited = new HashSet<>();
        q.add(S);
        int count = 0;
        while(!q.isEmpty()) {
            int size = q.size();
            for(int i = 0; i < size; i++) {
                int station = q.poll();

                if(station == T) return count;

                if(stationVisited.contains(station)) continue;

                stationVisited.add(station);
                for(int bus: stations.get(station)) {
                    if(busVisited.contains(bus)) continue;

                    busVisited.add(bus);
                    for(int newStation: routes[bus]) {
                        if(!stationVisited.contains(newStation)) q.offer(newStation);
                    }
                }
            }
            count++;
        }
        return -1;
    }
}
